lib/FIR/builder.h

```diff
@@ -15,106 +15,118 @@
 #ifndef FORTRAN_FIR_BUILDER_H_
 #define FORTRAN_FIR_BUILDER_H_
 
-#include "stmt.h"
+#include "statements.h"
 #include <initializer_list>
 
 namespace Fortran::FIR {
 
 struct FIRBuilder {
   explicit FIRBuilder(BasicBlock &block)
     : cursorRegion_{block.getParent()}, cursorBlock_{&block} {}
-  template<typename A> Statement &Insert(A &&s) {
+  template<typename A> Statement *Insert(A &&s) {
     CHECK(GetInsertionPoint());
     auto *statement{new Statement(GetInsertionPoint(), s)};
-    return *statement;
+    return statement;
```

Comment:
This should probably be `common::Indirection` instead of a raw pointer so that ownership is clear.

Reply:
And then `Indirection::Make()` can be used to create & construct the value.

Reply:
Ownership is not being passed to the caller. The idiom here is that a factory constructs FIR statements. Those statements will often need to be referenced to construct more FIR, and all the FIR will have a lifetime longer than the factory.

Reply:
Can you please clarify the ownership of the FIR nodes?  If it's not being granted to the caller of the factory, but the FIR will outlive the factory, who owns it?

Reply:
Ownership is managed exactly like LLVM, since the code reuses LLVM containers.

> Say you want to build a compilation unit. That's an object. You create the object and reference it with a pointer. The compilation unit has three functions and six global variables. The program owns those nine objects, and they are referenced by pointers. Function A has 12 basic blocks. Those are objects that are owned by the function and referenced by pointers. Basic block "entry" has 57 instructions. The basic block owns those instructions, which are referenced by pointers.

Reply:
That is a clear explanation.  Does this text or something similar appear in the FIR headers?  That would be great.

Reply:
I'll add it.